In [1]:
import mysql.connector
from sqlalchemy import create_engine,text
db_url = "mysql+pymysql://root:zeeshan786@localhost/project"
engine = create_engine(db_url)

In [2]:
import pandas as pd

In [3]:
query = "SHOW TABLES;"
tables = pd.read_sql(query, engine)
tables

Tables_in_project
0         customers
1         employees
2           offices
3      orderdetails
4            orders
5       paymentlogs
6          payments
7      productlines
8          products

In [4]:
## This helps us understand column names, data types, and relationships.

In [5]:
def describe_table(table_name):
    query = f"DESCRIBE {table_name};"
    df = pd.read_sql(query, engine)
    return df

# Check structure of all tables
tables_list = ["Customers", "Orders", "OrderDetails", "Products", 
               "Employees", "Offices", "Payments", "ProductLines"]

for table in tables_list:
    print(f"\n{table}")
    display(describe_table(table))


Customers


Field           Type Null  Key Default Extra
0           customerNumber            int   NO  PRI    None      
1             customerName    varchar(50)   NO         None      
2          contactLastName    varchar(50)   NO         None      
3         contactFirstName    varchar(50)   NO         None      
4                    phone    varchar(50)   NO         None      
5             addressLine1    varchar(50)   NO         None      
6             addressLine2    varchar(50)  YES         None      
7                     city    varchar(50)   NO         None      
8                    state    varchar(50)  YES         None      
9               postalCode    varchar(15)  YES         None      
10                 country    varchar(50)   NO         None      
11  salesRepEmployeeNumber            int  YES  MUL    None      
12             creditLimit  decimal(10,2)  YES         None


Orders


Field         Type Null  Key Default Extra
0     orderNumber          int   NO  PRI    None      
1       orderDate         date   NO  MUL    None      
2    requiredDate         date   NO         None      
3     shippedDate         date  YES         None      
4          status  varchar(15)   NO         None      
5        comments         text  YES         None      
6  customerNumber          int   NO  MUL    None


OrderDetails


Field           Type Null  Key Default Extra
0      orderNumber            int   NO  PRI    None      
1      productCode    varchar(15)   NO  PRI    None      
2  quantityOrdered            int   NO         None      
3        priceEach  decimal(10,2)   NO         None      
4  orderLineNumber       smallint   NO         None


Products


Field           Type Null  Key Default Extra
0         productCode    varchar(15)   NO  PRI    None      
1         productName    varchar(70)   NO         None      
2         productLine    varchar(50)   NO  MUL    None      
3        productScale    varchar(10)   NO         None      
4       productVendor    varchar(50)   NO         None      
5  productDescription           text   NO         None      
6     quantityInStock       smallint   NO         None      
7            buyPrice  decimal(10,2)   NO         None      
8                MSRP  decimal(10,2)   NO         None


Employees


Field          Type Null  Key Default Extra
0  employeeNumber           int   NO  PRI    None      
1        lastName   varchar(50)   NO         None      
2       firstName   varchar(50)   NO         None      
3       extension   varchar(10)   NO         None      
4           email  varchar(100)   NO         None      
5      officeCode   varchar(10)   NO  MUL    None      
6       reportsTo           int  YES  MUL    None      
7        jobTitle   varchar(50)   NO         None


Offices


Field         Type Null  Key Default Extra
0    officeCode  varchar(10)   NO  PRI    None      
1          city  varchar(50)   NO         None      
2         phone  varchar(50)   NO         None      
3  addressLine1  varchar(50)   NO         None      
4  addressLine2  varchar(50)  YES         None      
5         state  varchar(50)  YES         None      
6       country  varchar(50)   NO         None      
7    postalCode  varchar(15)   NO         None      
8     territory  varchar(10)   NO         None


Payments


Field           Type Null  Key Default Extra
0  customerNumber            int   NO  PRI    None      
1     checkNumber    varchar(50)   NO  PRI    None      
2     paymentDate           date   NO         None      
3          amount  decimal(10,2)   NO         None


ProductLines


Field           Type Null  Key Default Extra
0      productLine    varchar(50)   NO  PRI    None      
1  textDescription  varchar(4000)  YES         None      
2  htmlDescription     mediumtext  YES         None      
3            image     mediumblob  YES         None

## This tells us which tables contain the most data and which have fewer records.

In [7]:
query = '''
SELECT 'Customers' AS table_name, COUNT(*) AS total_rows FROM Customers
UNION ALL
SELECT 'Orders', COUNT(*) FROM Orders
UNION ALL
SELECT 'OrderDetails', COUNT(*) FROM OrderDetails
UNION ALL
SELECT 'Products', COUNT(*) FROM Products
UNION ALL
SELECT 'Employees', COUNT(*) FROM Employees
UNION ALL
SELECT 'Offices', COUNT(*) FROM Offices
UNION ALL
SELECT 'Payments', COUNT(*) FROM Payments
UNION ALL
SELECT 'ProductLines', COUNT(*) FROM ProductLines;
'''
row_count = pd.read_sql(query, engine)
row_count

table_name  total_rows
0     Customers         122
1        Orders         326
2  OrderDetails        2996
3      Products         110
4     Employees          23
5       Offices           7
6      Payments         273
7  ProductLines           7

In [8]:
## Preview Sample Data

In [9]:
def preview_table(table_name, limit=5):
    query = f"SELECT * FROM {table_name} LIMIT {limit};"
    df = pd.read_sql(query, engine)
    return df

# Preview each table
for table in tables_list:
    print(f"\n Table: {table}")
    display(preview_table(table))


 Table: Customers


customerNumber                customerName contactLastName  \
0             103           Atelier graphique         Schmitt   
1             112          Signal Gift Stores            King   
2             114  Australian Collectors, Co.        Ferguson   
3             119           La Rochelle Gifts         Labrune   
4             121          Baane Mini Imports      Bergulfsen   

  contactFirstName         phone                  addressLine1 addressLine2  \
0          Carine     40.32.2555                54, rue Royale         None   
1             Jean    7025551838               8489 Strong St.         None   
2            Peter  03 9520 4555             636 St Kilda Road      Level 3   
3          Janine     40.67.8555  67, rue des Cinquante Otages         None   
4           Jonas     07-98 9555        Erling Skakkes gate 78         None   

        city     state postalCode    country  salesRepEmployeeNumber  \
0     Nantes      None      44000     France                    1370   
1  Las Vegas        NV      83030        USA                    1166   
2  Melbourne  Victoria       3004  Australia                    1611   
3     Nantes      None      44000     France                    1370   
4    Stavern      None       4110     Norway                    1504   

   creditLimit  
0      21000.0  
1      71800.0  
2     117300.0  
3     118200.0  
4      81700.0


 Table: Orders


orderNumber   orderDate requiredDate shippedDate   status  \
0        10100  2003-01-06   2003-01-13  2003-01-10  Shipped   
1        10101  2003-01-09   2003-01-18  2003-01-11  Shipped   
2        10102  2003-01-10   2003-01-18  2003-01-14  Shipped   
3        10103  2003-01-29   2003-02-07  2003-02-02  Shipped   
4        10104  2003-01-31   2003-02-09  2003-02-01  Shipped   

                 comments  customerNumber  
0                    None             363  
1  Check on availability.             128  
2                    None             181  
3                    None             121  
4                    None             141


 Table: OrderDetails


orderNumber productCode  quantityOrdered  priceEach  orderLineNumber
0        10100    S18_1749               30     136.00                3
1        10100    S18_2248               50      55.09                2
2        10100    S18_4409               22      75.46                4
3        10100    S24_3969               49      35.29                1
4        10101    S18_2325               25     108.06                4


 Table: Products


productCode                            productName   productLine  \
0    S10_1678  1969 Harley Davidson Ultimate Chopper   Motorcycles   
1    S10_1949               1952 Alpine Renault 1300  Classic Cars   
2    S10_2016                  1996 Moto Guzzi 1100i   Motorcycles   
3    S10_4698   2003 Harley-Davidson Eagle Drag Bike   Motorcycles   
4    S10_4757                    1972 Alfa Romeo GTA  Classic Cars   

  productScale             productVendor  \
0         1:10           Min Lin Diecast   
1         1:10   Classic Metal Creations   
2         1:10  Highway 66 Mini Classics   
3         1:10         Red Start Diecast   
4         1:10   Motor City Art Classics   

                                  productDescription  quantityInStock  \
0  This replica features working kickstand, front...             7933   
1  Turnable front wheels; steering function; deta...             7305   
2  Official Moto Guzzi logos and insignias, saddl...             6625   
3  Model features, official Harley Davidson logos...             5582   
4  Features include: Turnable front wheels; steer...             3252   

   buyPrice    MSRP  
0     48.81   95.70  
1     98.58  214.30  
2     68.99  118.94  
3     91.02  193.66  
4     85.68  136.00


 Table: Employees


employeeNumber   lastName firstName extension  \
0            1002     Murphy     Diane     x5800   
1            1056  Patterson      Mary     x4611   
2            1076   Firrelli      Jeff     x9273   
3            1088  Patterson   William     x4871   
4            1102     Bondur    Gerard     x5408   

                             email officeCode  reportsTo              jobTitle  
0     dmurphy@classicmodelcars.com          1        NaN             President  
1   mpatterso@classicmodelcars.com          1     1002.0              VP Sales  
2   jfirrelli@classicmodelcars.com          1     1002.0          VP Marketing  
3  wpatterson@classicmodelcars.com          6     1056.0  Sales Manager (APAC)  
4     gbondur@classicmodelcars.com          4     1056.0   Sale Manager (EMEA)


 Table: Offices


officeCode           city            phone              addressLine1  \
0          1  San Francisco  +1 650 219 4782         100 Market Street   
1          2         Boston  +1 215 837 0825          1550 Court Place   
2          3            NYC  +1 212 555 3000      523 East 53rd Street   
3          4          Paris  +33 14 723 4404  43 Rue Jouffroy D'abbans   
4          5          Tokyo  +81 33 224 5000               4-1 Kioicho   

  addressLine2       state country postalCode territory  
0    Suite 300          CA     USA      94080        NA  
1    Suite 102          MA     USA      02107        NA  
2      apt. 5A          NY     USA      10022        NA  
3         None        None  France      75017      EMEA  
4         None  Chiyoda-Ku   Japan   102-8578     Japan


 Table: Payments


customerNumber checkNumber paymentDate    amount
0             103    HQ336336  2004-10-19   8066.78
1             103    JM555205  2003-06-05  14571.44
2             103    OM314933  2004-12-18   1676.14
3             112    BO864823  2004-12-17  14191.12
4             112     HQ55022  2003-06-06  32641.98


 Table: ProductLines


productLine                                    textDescription  \
0  Classic Cars  Attention car enthusiasts: Make your wildest c...   
1   Motorcycles  Our motorcycles are state of the art replicas ...   
2        Planes  Unique, diecast airplane and helicopter replic...   
3         Ships  The perfect holiday or anniversary gift for ex...   
4        Trains  Model trains are a rewarding hobby for enthusi...   

  htmlDescription image  
0            None  None  
1            None  None  
2            None  None  
3            None  None  
4            None  None

In [10]:
## This helps verify if the data looks clean or has missing/incorrect values.

## **Check for Missing Values**

In [12]:
def check_missing(table):
    query = f'SELECT SUM(null) AS missing_values FROM {table};'
    miss = pd.read_sql(query, engine)
    return miss
for table in tables_list:
    print(f'\n {table}')
    print(check_missing(table))


 Customers
  missing_values
0           None

 Orders
  missing_values
0           None

 OrderDetails
  missing_values
0           None

 Products
  missing_values
0           None

 Employees
  missing_values
0           None

 Offices
  missing_values
0           None

 Payments
  missing_values
0           None

 ProductLines
  missing_values
0           None


In [13]:
## there are no missing values in dataset because Missing values in key fields like customerNumber or orderNumber could cause analysis errors.

# lets check unique values for key columns

In [15]:
key_columns = {
    "Customers": ["customerNumber", "salesRepEmployeeNumber"],
    "Orders": ["orderNumber", "status"],
    "OrderDetails": ["orderNumber", "productCode"],
    "Products": ["productCode", "productLine"],
    "Employees": ["employeeNumber", "officeCode", "jobTitle"],
    "Payments": ["customerNumber", "checkNumber"],
    "ProductLines": ["productLine"],
}
for table, column in key_columns.items():
    for col in column:
        query = f'SELECT COUNT(DISTINCT {col}) FROM {table};'
        df = pd.read_sql(query, engine)
        print(f"{col}: {df.iloc[0, 0]} unique values")

customerNumber: 122 unique values
salesRepEmployeeNumber: 15 unique values
orderNumber: 326 unique values
status: 6 unique values
orderNumber: 326 unique values
productCode: 109 unique values
productCode: 110 unique values
productLine: 7 unique values
employeeNumber: 23 unique values
officeCode: 7 unique values
jobTitle: 7 unique values
customerNumber: 98 unique values
checkNumber: 273 unique values
productLine: 7 unique values


## check data types and value range

In [18]:
def data_summary(table_name):
    query = f"SELECT * FROM {table_name};"
    df = pd.read_sql(query, engine)
    return df.describe(include="all")

# Check summary for all tables
for table in tables_list:
    print(f"\n Summary for Table: {table}")
    display(data_summary(table))


 Summary for Table: Customers


customerNumber       customerName contactLastName contactFirstName  \
count       122.000000                122             122              122   
unique             NaN                122             108              111   
top                NaN  Atelier graphique           Young            Julie   
freq               NaN                  1               4                4   
mean        296.401639                NaN             NaN              NaN   
std         117.077544                NaN             NaN              NaN   
min         103.000000                NaN             NaN              NaN   
25%         191.250000                NaN             NaN              NaN   
50%         298.500000                NaN             NaN              NaN   
75%         384.250000                NaN             NaN              NaN   
max         496.000000                NaN             NaN              NaN   

             phone    addressLine1 addressLine2 city state postalCode country  \
count          122             122           22  122    49        115     122   
unique         121             122           21   96    18         94      28   
top     6175558555  54, rue Royale    Suite 101  NYC    CA      94217     USA   
freq             2               1            2    5    11          5      36   
mean           NaN             NaN          NaN  NaN   NaN        NaN     NaN   
std            NaN             NaN          NaN  NaN   NaN        NaN     NaN   
min            NaN             NaN          NaN  NaN   NaN        NaN     NaN   
25%            NaN             NaN          NaN  NaN   NaN        NaN     NaN   
50%            NaN             NaN          NaN  NaN   NaN        NaN     NaN   
75%            NaN             NaN          NaN  NaN   NaN        NaN     NaN   
max            NaN             NaN          NaN  NaN   NaN        NaN     NaN   

        salesRepEmployeeNumber    creditLimit  
count                100.00000     122.000000  
unique                     NaN            NaN  
top                        NaN            NaN  
freq                       NaN            NaN  
mean                1395.94000   67659.016393  
std                  165.67193   45043.370751  
min                 1165.00000       0.000000  
25%                 1286.00000   42175.000000  
50%                 1370.00000   76700.000000  
75%                 1504.00000   95075.000000  
max                 1702.00000  227600.000000


 Summary for Table: Orders


orderNumber   orderDate requiredDate shippedDate   status  \
count     326.000000         326          326         312      326   
unique           NaN         265          264         247        6   
top              NaN  2004-11-24   2004-12-01  2004-11-07  Shipped   
freq             NaN           4            5           4      303   
mean    10262.500000         NaN          NaN         NaN      NaN   
std        94.252321         NaN          NaN         NaN      NaN   
min     10100.000000         NaN          NaN         NaN      NaN   
25%     10181.250000         NaN          NaN         NaN      NaN   
50%     10262.500000         NaN          NaN         NaN      NaN   
75%     10343.750000         NaN          NaN         NaN      NaN   
max     10425.000000         NaN          NaN         NaN      NaN   

                                                 comments  customerNumber  
count                                                  80      326.000000  
unique                                                 37             NaN  
top     They want to reevaluate their terms agreement ...             NaN  
freq                                                    7             NaN  
mean                                                  NaN      263.834356  
std                                                   NaN      121.325792  
min                                                   NaN      103.000000  
25%                                                   NaN      145.250000  
50%                                                   NaN      240.000000  
75%                                                   NaN      360.750000  
max                                                   NaN      496.000000


 Summary for Table: OrderDetails


orderNumber productCode  quantityOrdered    priceEach  \
count    2996.000000        2996      2996.000000  2996.000000   
unique           NaN         109              NaN          NaN   
top              NaN    S18_3232              NaN          NaN   
freq             NaN          53              NaN          NaN   
mean    10260.352804         NaN        35.218959    90.769499   
std        92.476425         NaN         9.833885    36.582917   
min     10100.000000         NaN         6.000000    26.550000   
25%     10181.000000         NaN        27.000000    62.000000   
50%     10262.000000         NaN        35.000000    85.805000   
75%     10338.250000         NaN        43.000000   114.650000   
max     10425.000000         NaN        97.000000   214.300000   

        orderLineNumber  
count       2996.000000  
unique              NaN  
top                 NaN  
freq                NaN  
mean           6.430574  
std            4.197167  
min            1.000000  
25%            3.000000  
50%            6.000000  
75%            9.000000  
max           18.000000


 Summary for Table: Products


productCode                            productName   productLine  \
count          110                                    110           110   
unique         110                                    110             7   
top       S10_1678  1969 Harley Davidson Ultimate Chopper  Classic Cars   
freq             1                                      1            38   
mean           NaN                                    NaN           NaN   
std            NaN                                    NaN           NaN   
min            NaN                                    NaN           NaN   
25%            NaN                                    NaN           NaN   
50%            NaN                                    NaN           NaN   
75%            NaN                                    NaN           NaN   
max            NaN                                    NaN           NaN   

       productScale            productVendor  \
count           110                      110   
unique            8                       13   
top            1:18  Classic Metal Creations   
freq             42                       10   
mean            NaN                      NaN   
std             NaN                      NaN   
min             NaN                      NaN   
25%             NaN                      NaN   
50%             NaN                      NaN   
75%             NaN                      NaN   
max             NaN                      NaN   

                                       productDescription  quantityInStock  \
count                                                 110       110.000000   
unique                                                 95              NaN   
top     Turnable front wheels; steering function; deta...              NaN   
freq                                                    4              NaN   
mean                                                  NaN      5046.645455   
std                                                   NaN      2939.263668   
min                                                   NaN        15.000000   
25%                                                   NaN      2419.000000   
50%                                                   NaN      5437.500000   
75%                                                   NaN      7599.750000   
max                                                   NaN      9997.000000   

          buyPrice        MSRP  
count   110.000000  110.000000  
unique         NaN         NaN  
top            NaN         NaN  
freq           NaN         NaN  
mean     54.395182  100.438727  
std      22.051130   39.663649  
min      15.910000   33.190000  
25%      34.180000   68.907500  
50%      53.915000   98.300000  
75%      68.675000  122.850000  
max     103.420000  214.300000


 Summary for Table: Employees


employeeNumber   lastName firstName extension  \
count        23.000000         23        23        23   
unique             NaN         19        21        20   
top                NaN  Patterson    Gerard      x102   
freq               NaN          3         2         3   
mean       1335.391304        NaN       NaN       NaN   
std         223.565475        NaN       NaN       NaN   
min        1002.000000        NaN       NaN       NaN   
25%        1154.000000        NaN       NaN       NaN   
50%        1323.000000        NaN       NaN       NaN   
75%        1557.500000        NaN       NaN       NaN   
max        1702.000000        NaN       NaN       NaN   

                                 email officeCode    reportsTo   jobTitle  
count                               23         23    22.000000         23  
unique                              22          7          NaN          7  
top     jfirrelli@classicmodelcars.com          1          NaN  Sales Rep  
freq                                 2          6          NaN         17  
mean                               NaN        NaN  1117.409091        NaN  
std                                NaN        NaN   120.173549        NaN  
min                                NaN        NaN  1002.000000        NaN  
25%                                NaN        NaN  1064.000000        NaN  
50%                                NaN        NaN  1102.000000        NaN  
75%                                NaN        NaN  1143.000000        NaN  
max                                NaN        NaN  1621.000000        NaN


 Summary for Table: Offices


officeCode           city            phone       addressLine1  \
count           7              7                7                  7   
unique          7              7                7                  7   
top             1  San Francisco  +1 650 219 4782  100 Market Street   
freq            1              1                1                  1   

       addressLine2 state country postalCode territory  
count             5     4       7          7         7  
unique            5     4       5          7         4  
top       Suite 300    CA     USA      94080        NA  
freq              1     1       3          1         3


 Summary for Table: Payments


customerNumber checkNumber paymentDate         amount
count       273.000000         273         273     273.000000
unique             NaN         273         232            NaN
top                NaN    HQ336336  2004-06-21            NaN
freq               NaN           1           3            NaN
mean        271.194139         NaN         NaN   32438.971538
std         120.069507         NaN         NaN   20988.231279
min         103.000000         NaN         NaN     615.450000
25%         161.000000         NaN         NaN   16909.840000
50%         250.000000         NaN         NaN   32077.440000
75%         363.000000         NaN         NaN   45352.470000
max         496.000000         NaN         NaN  120166.580000


 Summary for Table: ProductLines


productLine                                    textDescription  \
count              7                                                  7   
unique             7                                                  7   
top     Classic Cars  Attention car enthusiasts: Make your wildest c...   
freq               1                                                  1   

       htmlDescription image  
count                0     0  
unique               0     0  
top                NaN   NaN  
freq               NaN   NaN

In [19]:
## almost everything is correct

# Check Duplicate Entries

In [21]:
query = """
SELECT customerNumber, COUNT(*) 
FROM Customers 
GROUP BY customerNumber 
HAVING COUNT(*) > 1;
"""
duplicates = pd.read_sql(query, engine)
display(duplicates)

Empty DataFrame
Columns: [customerNumber, COUNT(*)]
Index: []

In [22]:
# there is no duplicate exists

In [23]:
## Total Customers

In [24]:
query = '''SELECT COUNT(*) FROM Customers;'''
customer_count = pd.read_sql(query, engine)
display(customer_count)

COUNT(*)
0       122

### Insight on Customer Count (122 Customers)  
The database contains **122 unique customers**, which indicates the size of the customer base. This metric is important for business growth analysis.  

- If the number of customers is **low**, the company may need to focus on **customer acquisition strategies**.  
- If the number is **high**, the company should analyze **customer retention and repeat orders**.  

In [26]:
query = '''SELECT c.customerNumber, c.contactFirstName, SUM(p.amount) AS total_payments
FROM Customers c JOIN Payments p ON c.customerNumber = p.customerNumber 
GROUP BY c.customerNumber, c.contactFirstName
ORDER BY total_payments DESC LIMIT 5;
'''
top_five_customers = pd.read_sql(query, engine)
display(top_five_customers)

customerNumber contactFirstName  total_payments
0             141           Diego        715738.98
1             124            Susan       584188.24
2             114            Peter       180585.07
3             151             Jeff       177913.95
4             148             Eric       156251.03

### Insight on Top 5 Paying Customers  

The top 5 customers who have made the **highest total payments** are:  

1. **Diego** - **$715,738.98**  

2. **Susan** - **$584,188.24**  

3. **Peter** - **$180,585.07**  

4. **Jeff** - **$177,913.95**  

5. **Eric** - **$156,251.03**  

### Key Observations:  
- **Diego and Susan** are the **top two high-value customers**, contributing significantly to the company's revenue.  
- The gap between the **top two customers and the rest** is substantial, suggesting a few high-value clients dominate sales.  

In [117]:
query = '''SELECT country, count(*) AS num_of_customers FROM Customers GROUP BY country ORDER BY num_of_customers DESC'''
customer_by_country = pd.read_sql(query, engine)
display(customer_by_country.head())

country  num_of_customers
0        USA                36
1    Germany                13
2     France                12
3      Spain                 7
4  Australia                 5

### Insights:
- **USA** has the highest number of customers (36), making it the largest market.
- **France** (12) and **Germany** (13) also have strong customer bases.
- Several countries, including **Poland**, **Hong Kong**, and **Russia**, have only one customer, indicating low presence.
- The data suggests potential market expansion opportunities in low-representation countries.

## inactive customers

In [30]:
query = 'SELECT customerNumber FROM Customers WHERE customerNumber NOT IN(SELECT customerNumber FROM Orders);'
customer_without_order = pd.read_sql(query, engine)
display(customer_without_order.head())

customerNumber
0             125
1             169
2             206
3             223
4             237

## Insights from This Query  

**Identifies inactive customers** → Helps analyze why they haven't placed an order.  

# Customers with their sales representative

In [33]:
query = '''
SELECT CONCAT(c.contactFirstName, ' ', contactLastName) AS customer_name, CONCAT(e.firstName, ' ', lastName) AS Sales_representative
FROM Customers c JOIN Employees e ON c.salesRepEmployeeNumber = e.employeeNumber
WHERE c.salesRepEmployeeNumber IN(SELECT employeeNumber FROM Employees);
'''
customer_saler = pd.read_sql(query, engine)
display(customer_saler.head())

customer_name Sales_representative
0    Susan Nelson      Leslie Jennings
1    Julie Murphy      Leslie Jennings
2  Juri Hashimoto      Leslie Jennings
3     Julie Brown      Leslie Jennings
4       Sue Frick      Leslie Jennings

## Insights from This Query  

1️ Helps in understanding which sales rep handles which customers.  

2️ Can be used to analyze the **performance of different sales reps**.  

In [35]:
query = '''SELECT c.customerNumber, COUNT(*) AS num_of_orders FROM Customers c JOIN Orders o ON c.customerNumber = o.customerNumber
GROUP BY c.customerNumber order by num_of_orders DESC;
'''
customer_with_orders = pd.read_sql(query, engine)
display(customer_with_orders.head())

customerNumber  num_of_orders
0             141             26
1             124             17
2             114              5
3             353              5
4             145              5

## Insights  

- Customer **141** stands out with the highest number of orders (**26**), indicating strong engagement.  
- Customers like **124** and **114** have also made significant purchases, with **17** and **5** orders, respectively.  
- Multiple customers with **5 orders** suggest a group of frequent buyers who can be targeted for loyalty programs.  
- This data helps prioritize top customers for personalized offers and inventory adjustments.  

## Total Revenue

In [38]:
query = 'SELECT SUM(amount) AS Total_Revenue FROM Payments;'
total_revenue = pd.read_sql(query, engine)
display(total_revenue)

Total_Revenue
0     8855839.23

## Insights  

- The **total revenue** generated from all payments is **8853839.23**, reflecting the overall earnings of the business.   

# Top selling Products

In [40]:
query = '''SELECT p.productName, COUNT(od.productCode) AS count
FROM Products p
JOIN OrderDetails od ON p.productCode = od.productCode
GROUP BY p.productName
ORDER BY count DESC;
'''
frequent_order_products = pd.read_sql(query, engine)
frequent_order_products.head()

productName  count
0            1992 Ferrari 360 Spider red     53
1  1969 Harley Davidson Ultimate Chopper     28
2               1952 Alpine Renault 1300     28
3                  1996 Moto Guzzi 1100i     28
4   2003 Harley-Davidson Eagle Drag Bike     28

## Insights  

- The **1992 Ferrari 360 Spider (red)** is the most frequently ordered product with **53 orders**, indicating high demand.  
- Other products, such as **1969 Harley Davidson Ultimate Chopper** and **1952 Alpine Renault 1300**, have **28 orders each**, making them popular choices as well.  
- Identifying top-selling products helps in **inventory management** and **restocking strategies** to ensure availability.  

# Order count by Status

In [42]:
query = '''SELECT status, COUNT(orderNumber) AS order_count FROM Orders
GROUP BY status ORDER BY order_count DESC;
'''
order_count_by_status = pd.read_sql(query, engine)
display(order_count_by_status)

status  order_count
0     Shipped          303
1   Cancelled            6
2  In Process            6
3    Resolved            4
4     On Hold            4
5    Disputed            3

## Insights  

- The majority of orders (**303**) are **shipped**, showing efficient order fulfillment.  
- **6 orders** are **cancelled** and **6 are in process**, which may require review to prevent further delays.  
- **4 orders** are marked as **resolved**, possibly after disputes or issues.  
- **4 orders** are **on hold**, which could indicate stock issues or pending approvals.  
- **3 orders** are **disputed**, suggesting customer complaints or transaction issues that need resolution.  

## Least Ordererd Products

In [45]:
query = '''SELECT p.productName, COUNT(od.productCode) AS count
FROM Products p
JOIN OrderDetails od ON p.productCode = od.productCode
GROUP BY p.productName
ORDER BY count ASC;
'''
frequent_order_products = pd.read_sql(query, engine)
frequent_order_products.head()

productName  count
0     1957 Ford Thunderbird     24
1         1952 Citroen-15CV     24
2     1965 Aston Martin DB5     25
3  1917 Grand Touring Sedan     25
4        1911 Ford Town Car     25

## Insights  

- The **1957 Ford Thunderbird** and **1952 Citroen-15CV** are the least ordered products, with only **24 orders** each.  
- Other products, like the **1965 Aston Martin DB5**, **1917 Grand Touring Sedan**, and **1911 Ford Town Car**, also have low order counts (25 each).  
- These low-demand products might indicate a **lack of customer interest**, **pricing issues**, or **ineffective marketing**.  

## Customer who made orders but not give payment

In [48]:
query = '''
SELECT c.customerNumber, p.amount FROM Customers c JOIN Payments p ON c.customerNumber = p.customerNumber
WHERE c.customerNumber IN(SELECT customerNumber FROM Orders)
AND c.customerNumber IN(SELECT customerNumber FROM Payments WHERE amount = 0);
'''
without_payment = pd.read_sql(query, engine)
display(without_payment.head())

Empty DataFrame
Columns: [customerNumber, amount]
Index: []

## Insights  

- No customers were found with **orders placed but no payments made**, meaning all orders have been properly accounted for.  
- This indicates **efficient payment processing**, minimizing issues related to **unpaid invoices** or **outstanding balances**.  
- The business has a **healthy cash flow**, as all customers making purchases are also completing their payments.  

# Customer who gives payment but not made order

In [50]:
query = '''
SELECT c.customerNumber, p.amount FROM Customers c JOIN Payments p ON c.customerNumber = p.customerNumber
WHERE c.customerNumber NOT IN(SELECT customerNumber FROM Orders)
AND c.customerNumber IN(SELECT customerNumber FROM Payments WHERE amount IS NOT NULL);
'''
without_order = pd.read_sql(query, engine)
display(without_order.head())

Empty DataFrame
Columns: [customerNumber, amount]
Index: []

## Insights
All payments belong to customers who have placed orders.

## Finding AOV (Average Order Value)

In [53]:
query = '''
SELECT (SELECT SUM(amount) AS Revenue FROM Payments) /
(SELECT COUNT(orderNumber) AS total_orders FROM Orders) AS AOV
'''
AVO = pd.read_sql(query, engine)
display(AVO)

AOV
0  27165.151012

## Insights  

- The **Average Order Value (AOV)** is approximately **$27,159.02**.  
- This indicates that, on average, each order generates a substantial revenue, suggesting **high-value transactions**.  

## Total SUM of Order Values Per Customer

In [56]:
query = '''
SELECT CONCAT(c.contactFirstName, ' ', c.contactLastName) AS Customer_Name, SUM(p.amount) AS total_amount FROM Customers c
JOIN Payments p ON c.customerNumber = p.customerNumber GROUP BY Customer_Name ORDER BY total_amount DESC;
'''
highest_order_value = pd.read_sql(query, engine)
display(highest_order_value.head())

Customer_Name  total_amount
0   Diego  Freyre     715738.98
1    Susan Nelson     584188.24
2  Peter Ferguson     180585.07
3      Jeff Young     177913.95
4  Eric Natividad     156251.03

## Insights  

- **Diego Freyre** is the highest spender with a total payment of 715,738.98, followed by  Susan Nelson  at 584,188.24.  
- The top 5 customers have significantly high payments, indicating that they are **high-value clients** and **key contributors** to the business's revenue.  

## Product's with the Highest Revenue

In [59]:
query = '''SELECT p.productName, SUM(py.amount) as Revenue FROM Products p 
JOIN OrderDetails od ON p.productCode = od.productCode 
JOIN Orders o ON od.orderNumber = o.orderNumber
JOIN Customers c ON o.customerNumber = c.customerNumber
JOIN Payments py ON c.customerNumber = py.customerNumber
GROUP BY p.productName ORDER BY Revenue DESC;
'''
highest_revenue_prod = pd.read_sql(query, engine)
highest_revenue_prod.head()

productName      Revenue
0          1992 Ferrari 360 Spider red  10842745.04
1  1958 Chevy Corvette Limited Edition   8428983.28
2                   1970 Dodge Coronet   7867836.37
3          1954 Greyhound Scenicruiser   7852167.03
4     Diamond T620 Semi-Skirted Tanker   7241068.96

## Insights

- **1992 Ferrari 360 Spider red** is the highest-grossing product with a revenue of 10,842,745.04, followed by the 1958 Chevy Corvette Limited Edition at 8,428,983.28.  
- The top products indicate that **luxury/high-end products** contribute significantly to overall revenue.  
- Businesses can focus on promoting and ensuring stock availability for high-revenue products.  

## Revenue By Product Line

In [62]:
query = '''
SELECT p.productline AS Product_line, SUM(py.amount) AS Revenue FROM Products p
JOIN OrderDetails od ON p.productCode = od.productCode
JOIN Orders o ON od.orderNumber = o.orderNumber 
JOIN Customers c ON o.customerNumber = c.customerNumber
JOIN Payments py ON c.customerNumber = py.customerNumber
GROUP BY Product_line ORDER BY Revenue DESC;
'''
product_line_revenue = pd.read_sql(query, engine)
display(product_line_revenue.head())

Product_line       Revenue
0      Classic Cars  1.957233e+08
1      Vintage Cars  1.120428e+08
2  Trucks and Buses  6.544744e+07
3       Motorcycles  4.828474e+07
4            Planes  4.417167e+07

## Insights  

- **Classic Cars** contribute the highest revenue (195.72M), making them the most profitable product line.  
- **Vintage Cars** rank second with a total revenue of (112.04M), indicating strong demand.  
- **Trucks and Buses** generate (65.45M), securing third place in revenue contribution.  
- **Motorcycles** and **Planes** generate revenues of (48.28M) and **$44.17M**, respectively.  
- This data helps businesses **focus on high-performing product lines**.  

## Month Wise Revenue

In [65]:
query = '''
SELECT MONTH(o.orderDate) AS Month, SUM(py.amount) AS Revenue FROM orders o
JOIN customers c ON o.customerNumber = c.customerNumber 
JOIN Payments py ON c.customerNumber = py.customerNumber
GROUP BY Month ORDER BY Revenue DESC
'''
monthly_revenue = pd.read_sql(query, engine)
display(monthly_revenue)

Month     Revenue
0      11  6970568.02
1       2  5770507.95
2      12  5574473.45
3       3  5573687.24
4       5  5395712.43
5       1  4392053.85
6      10  4353071.22
7       6  3556045.65
8       8  3107394.80
9       4  3069367.36
10      7  2641845.24
11      9  2107349.16

## Insights  

- **November** generated the highest revenue (**6,968,568.02**), indicating it might be a peak sales period, possibly due to holiday shopping.  
- **February and December** also had strong revenues, suggesting seasonal trends affecting customer spending.  
- The **lowest revenue month was September (2,105,349.16)**, indicating a potential slow period for sales.  

## Year Wise Revenue

In [68]:
query = '''
SELECT YEAR(o.orderDate) AS Year, SUM(py.amount) AS Revenue FROM orders o
JOIN customers c ON o.customerNumber = c.customerNumber 
JOIN Payments py ON c.customerNumber = py.customerNumber
GROUP BY Year ORDER BY Revenue DESC
'''
year_revenue = pd.read_sql(query, engine)
display(year_revenue)

Year      Revenue
0  2004  21108232.41
1  2003  16617208.41
2  2005  14786635.55

## Insights  

- **2004** had the highest revenue (**21,104,232.41**), indicating it was the most profitable year.  
- **2003** and **2005** followed with **16,615,208.41** and **14,786,635.55**, showing a decline in revenue over time.  
- The decreasing trend from **2004 to 2005** might indicate market saturation, economic factors, or shifts in customer demand.  

## Ranking Customers by Total Spending

In [73]:
query = '''WITH CustomerRevenue AS (
    SELECT c.customerNumber, CONCAT(c.contactFirstName, ' ', c.contactLastName) AS Customer_Name, SUM(py.amount) AS Total_Revenue
    FROM Customers c
    JOIN Payments py ON c.customerNumber = py.customerNumber
    GROUP BY c.customerNumber
)
SELECT Customer_Name, Total_Revenue, RANK() OVER (ORDER BY Total_Revenue DESC) AS 'Rank'
FROM CustomerRevenue;'''
ranking_by_spending = pd.read_sql(query, engine)
display(ranking_by_spending.head())

Customer_Name  Total_Revenue  Rank
0   Diego  Freyre      715738.98     1
1    Susan Nelson      584188.24     2
2  Peter Ferguson      180585.07     3
3      Jeff Young      177913.95     4
4  Eric Natividad      156251.03     5

## Highest Spending Customer in Each Year

In [75]:
query = '''SELECT 
    YEAR(o.orderDate) AS Year,
    c.customerNumber,
    CONCAT(c.contactFirstName, ' ', c.contactLastName) AS Customer_Name,
    SUM(py.amount) AS Total_Spend
FROM Orders o
JOIN Customers c ON o.customerNumber = c.customerNumber
JOIN Payments py ON c.customerNumber = py.customerNumber
WHERE YEAR(o.orderDate) = (SELECT MAX(YEAR(o2.orderDate)) FROM Orders o2) 
GROUP BY YEAR(o.orderDate), c.customerNumber
ORDER BY Total_Spend DESC
LIMIT 1;'''
pd.read_sql(query, engine)

Year  customerNumber  Customer_Name  Total_Spend
0  2005             141  Diego  Freyre   6441650.82

## Year-over-Year Growth in Revenue for Each Customer

In [77]:
query = '''WITH YearlyRevenue AS (
  SELECT 
    YEAR(o.orderDate) AS Year,
    o.customerNumber,
    SUM(py.amount) AS Revenue
  FROM Orders o
  JOIN Customers c ON o.customerNumber = c.customerNumber
  JOIN Payments py ON c.customerNumber = py.customerNumber
  GROUP BY YEAR(o.orderDate), o.customerNumber
)
SELECT 
  y1.customerNumber,
  y1.Year,
  y1.Revenue AS Current_Year_Revenue,
  y2.Revenue AS Previous_Year_Revenue,
  ((y1.Revenue - y2.Revenue) / y2.Revenue) * 100 AS Revenue_Growth_Percentage
FROM YearlyRevenue y1
LEFT JOIN YearlyRevenue y2 ON y1.customerNumber = y2.customerNumber AND y1.Year = y2.Year + 1
ORDER BY Revenue_Growth_Percentage DESC;'''
pd.read_sql(query, engine).head()

customerNumber  Year  Current_Year_Revenue  Previous_Year_Revenue  \
0             145  2004             322339.50              107446.50   
1             131  2004             322919.82              107639.94   
2             144  2004             131041.95               43680.65   
3             146  2004             260610.70              130305.35   
4             458  2004             224880.18              112440.09   

   Revenue_Growth_Percentage  
0                      200.0  
1                      200.0  
2                      200.0  
3                      100.0  
4                      100.0

## Customers who have not made a purchase in the Last 6 Months

In [139]:
query = '''SELECT 
  c.customerNumber,
  CONCAT(c.contactFirstName, ' ', c.contactLastName) AS Customer_Name
FROM Customers c
LEFT JOIN Orders o ON c.customerNumber = o.customerNumber
WHERE o.orderDate < DATE_SUB(CURDATE(), INTERVAL 6 MONTH)
GROUP BY c.customerNumber
HAVING COUNT(o.orderNumber) = 0;
'''
pd.read_sql(query, engine)

Empty DataFrame
Columns: [customerNumber, Customer_Name]
Index: []

## Revenue Distribution for Each Product Line

In [154]:
query = '''WITH RankedProducts AS (
  SELECT 
    p.productLine,
    p.productName,
    SUM(py.amount) AS Revenue,
    RANK() OVER (PARTITION BY p.productLine ORDER BY SUM(py.amount) DESC) AS Product_Rank
  FROM Products p
  JOIN OrderDetails od ON p.productCode = od.productCode
  JOIN Orders o ON od.orderNumber = o.orderNumber
  JOIN Customers c ON o.customerNumber = c.customerNumber
  JOIN Payments py ON c.customerNumber = py.customerNumber
  GROUP BY p.productLine, p.productName
)
SELECT productLine, productName, Revenue
FROM RankedProducts
WHERE Product_Rank <= 5
ORDER BY productLine, Revenue DESC;'''
pd.read_sql(query, engine).sample(5)

productLine                        productName     Revenue
4       Classic Cars  1992 Porsche Cayenne Turbo Silver  7145099.86
6        Motorcycles            1982 Ducati 900 Monster  4532986.21
24  Trucks and Buses   Diamond T620 Semi-Skirted Tanker  7241068.96
31      Vintage Cars                1904 Buick Runabout  5884885.33
12            Planes                 F/A 18 Hornet 1/72  3863320.86

## (AOV) per Customer Segmented by Year

In [156]:
query = '''
WITH CustomerOrders AS (
    SELECT 
    c.customerNumber,
    YEAR(o.orderDate) AS Year,
    SUM(py.amount) AS Total_Revenue,
    COUNT(o.orderNumber) AS Total_Orders
  FROM Orders o
  JOIN Customers c ON o.customerNumber = c.customerNumber
  JOIN Payments py ON c.customerNumber = py.customerNumber
  GROUP BY c.customerNumber, YEAR(o.orderDate)
)
SELECT 
  customerNumber,
  Year,
  Total_Revenue / Total_Orders AS Average_Order_Value
FROM CustomerOrders
ORDER BY Year DESC, Average_Order_Value DESC;
'''
pd.read_sql(query, engine).sample(6)

customerNumber  Year  Average_Order_Value
191             129  2003         22236.853333
184             350  2003         23849.176667
82              353  2004         31745.797500
36              471  2005         22460.380000
9               448  2005         38388.220000
147             148  2003         39062.757500

## INDEXING

# Non-Clustered Index on orderDate in the Orders Table

In [160]:
with engine.connect() as connection:
    connection.execute(text("CREATE INDEX idx_order_date ON Orders(orderDate);"))
    connection.commit()

In [86]:
query = text("SHOW INDEX FROM Orders;")
df_indexes = pd.read_sql(query, engine)
display(df_indexes)

Table  Non_unique              Key_name  Seq_in_index     Column_name  \
0  orders           0               PRIMARY             1     orderNumber   
1  orders           1        customerNumber             1  customerNumber   
2  orders           1           index_order             1  customerNumber   
3  orders           1  idx_orderDate_status             1       orderDate   
4  orders           1  idx_orderDate_status             2          status   

  Collation  Cardinality Sub_part Packed Null Index_type Comment  \
0         A          326     None   None           BTREE           
1         A           98     None   None           BTREE           
2         A           98     None   None           BTREE           
3         A          265     None   None           BTREE           
4         A          269     None   None           BTREE           

  Index_comment Visible Expression  
0                   YES       None  
1                   YES       None  
2                   YES       None  
3                   YES       None  
4                   YES       None

# Composite index

In [171]:
with engine.connect() as conn:
    conn.execute(text("CREATE INDEX idx_orderDate_status ON Orders(orderDate, status);"))
    conn.commit()

### TRANSACTIONS

In [100]:
with engine.begin() as conn:
    conn.execute(text("""
        CREATE TABLE IF NOT EXISTS PaymentLogs (
            logID INT AUTO_INCREMENT PRIMARY KEY,
            customerNumber INT,
            checkNumber VARCHAR(50),
            action TEXT,
            logDate DATETIME,
            FOREIGN KEY (customerNumber) REFERENCES Customers(customerNumber)
        );
    """))

In [102]:
with engine.begin() as conn:
    conn.execute(text("""
        UPDATE Payments
        SET amount = amount + 1000
        WHERE customerNumber = 103 AND checkNumber = 'HQ336336'
    """))

    conn.execute(text("""
        INSERT INTO PaymentLogs (customerNumber, checkNumber, action, logDate)
        VALUES (103, 'HQ336336', 'Corrected amount by +1000', NOW())
    """))

In [104]:
query = '''
SELECT customerNumber, checkNumber, amount
FROM Payments
WHERE customerNumber = 103 AND checkNumber = 'HQ336336';
'''
pd.read_sql(query, engine)

customerNumber checkNumber   amount
0             103    HQ336336  9066.78

In [106]:
query = '''
SELECT * FROM PaymentLogs
WHERE customerNumber = 103 AND checkNumber = 'HQ336336'
ORDER BY logDate DESC
LIMIT 1;
'''
pd.read_sql(query, engine)

logID  customerNumber checkNumber                     action  \
0      4             103    HQ336336  Corrected amount by +1000   

              logDate  
0 2025-04-04 06:25:15

#### A transaction was used to update the payment amount for customer 103 and simultaneously log the action in a separate table PaymentLogs.
The new payment amount is $8066.78.

In [239]:
from sqlalchemy.orm import sessionmaker
Session = sessionmaker(bind=engine)
session = Session()
try:
    session.begin()
    session.execute("""
        UPDATE Payments
        SET amount = amount + 1000
        WHERE customerNumber = 103 AND checkNumber = 'HQ336336';
    """)
    session.execute("""
        INSERT INTO Payments (customerNumber, checkNumber, paymentDate, amount)
        VALUES (9999, 'HQ336337', '2025-04-05', 500);
    """)
    session.commit()
except Exception as e:
    print(f"An error occurred: {e}")
    session.rollback()
finally:
    session.close()

An error occurred: Textual SQL expression '\n        UPDATE Payments\n...' should be explicitly declared as text('\n        UPDATE Payments\n...')


# Stored Procedure

### Get Customer's Total Payments through customerNumber

In [185]:
with engine.begin() as conn:
    conn.execute(text('''
    CREATE PROCEDURE GetCustomerTotalPayment(IN cust_num INT)
    BEGIN
        SELECT customerNumber, SUM(amount) AS Total_Payment
        FROM Payments
        WHERE customerNumber = cust_num
        GROUP BY customerNumber;
    END;'''))

In [187]:
query = "CALL GetCustomerTotalPayment(:cust_num);"
result = pd.read_sql(text(query), engine, params={"cust_num": 103})

display(result)

customerNumber  Total_Payment
0             103       25314.36

### Stored Procedure: `GetCustomerTotalPayment`

This stored procedure is designed to **retrieve the total payment made by a specific customer** using their `customerNumber`.  
It encapsulates a `SELECT` query that **sums up the payment amounts** from the `Payments` table for the given customer, returning:

- `customerNumber`
- `Total_Payment`

#### Real-World Use Case:
> In an e-commerce system, this can be used to **quickly retrieve total customer spending**, which is essential for loyalty programs, billing summaries, or fraud detection.

## Lets create an View for customer orders and payment

In [207]:
with engine.begin() as conn:
    conn.execute(text('''
        CREATE OR REPLACE VIEW CustomerSummary AS
        SELECT 
            c.customerNumber,
            CONCAT(c.contactFirstName,' ', c.contactLastName) AS customer_name,
            COUNT(DISTINCT o.orderNumber) AS total_orders,
            SUM(p.amount) AS total_payments
        FROM Customers c
        LEFT JOIN Orders o ON c.customerNumber = o.customerNumber
        LEFT JOIN Payments p ON c.customerNumber = p.customerNumber
        GROUP BY c.customerNumber, c.customerName
    '''))

In [209]:
query = "SELECT * FROM CustomerSummary ORDER BY total_payments DESC LIMIT 5"
top_customers = pd.read_sql(query, engine)
display(top_customers)

customerNumber   customer_name  total_orders  total_payments
0             141   Diego  Freyre            26     18609213.48
1             124    Susan Nelson            17      9931200.08
2             114  Peter Ferguson             5       902925.35
3             148  Eric Natividad             5       781255.15
4             323     Mike Graham             5       773110.40

### Using Views

I created a **view** called `CustomerSummary` that provides:
- Each customer's name.
- Total number of orders.
- Total payment amount.

This view simplifies repeated analytical queries by combining multiple joins and aggregations into a reusable, virtual table.

#### Benefits:
- Improved readability for complex joins.
- Simplifies reporting and analytics.
- Can be queried like a regular table.

## Complex Joins with Multiple Tables

In [219]:
query = '''
SELECT 
    c.customerNumber,
    CONCAT(c.contactFirstName,' ',c.contactLastName) AS Customer_Name,
    p.productName AS Product_Name,
    SUM(od.quantityOrdered) AS Total_Quantity_Ordered
FROM Customers c
JOIN Orders o ON c.customerNumber = o.customerNumber
JOIN OrderDetails od ON o.orderNumber = od.orderNumber
JOIN Products p ON od.productCode = p.productCode
GROUP BY c.customerNumber, p.productCode
ORDER BY Total_Quantity_Ordered DESC
LIMIT 10;
'''
pd.read_sql(query, engine)

customerNumber  Customer_Name                         Product_Name  \
0             141  Diego  Freyre          1992 Ferrari 360 Spider red   
1             141  Diego  Freyre  1958 Chevy Corvette Limited Edition   
2             141  Diego  Freyre                   1970 Dodge Coronet   
3             141  Diego  Freyre                    1957 Chevy Pickup   
4             141  Diego  Freyre                  2002 Chevy Corvette   
5             124   Susan Nelson              1939 Cadillac Limousine   
6             141  Diego  Freyre              1956 Porsche 356A Coupe   
7             141  Diego  Freyre    1992 Porsche Cayenne Turbo Silver   
8             141  Diego  Freyre                   1949 Jaguar XK 120   
9             141  Diego  Freyre          1954 Greyhound Scenicruiser   

   Total_Quantity_Ordered  
0                   308.0  
1                   245.0  
2                   197.0  
3                   183.0  
4                   174.0  
5                   167.0  
6                   161.0  
7                   161.0  
8                   158.0  
9                   157.0

### Complex Joins with Multiple Tables

This query combines **four tables**: `Customers`, `Orders`, `OrderDetails`, and `Products`, to find the **top products** ordered by **customers** and the **total quantity** of each product ordered.

---

#### Key Points:
- **Joins** are fundamental for **combining data** from multiple tables in relational databases.
- Using **aggregate functions** like `SUM()` helps to derive meaningful insights from raw transactional data.

## Year wise order status with number_of_orders and total_order_amount

In [230]:
query = """
SELECT 
    YEAR(o.orderDate) AS order_year,
    o.status AS order_status,
    COUNT(o.orderNumber) AS number_of_orders,
    SUM(od.quantityOrdered * od.priceEach) AS total_order_amount
FROM Orders o
JOIN OrderDetails od ON o.orderNumber = od.orderNumber
GROUP BY Order_Year, o.status
ORDER BY Order_Year, Total_Order_Amount DESC;
"""
pd.read_sql(query, engine)

order_year order_status  number_of_orders  total_order_amount
0         2003      Shipped              1019          3223095.80
1         2003    Cancelled                25            67130.69
2         2003     Resolved                 8            27121.90
3         2004      Shipped              1353          4300602.99
4         2004    Cancelled                54           171723.49
5         2004      On Hold                 6            23014.17
6         2004     Resolved                 8            20564.86
7         2005      Shipped               399          1341395.85
8         2005      On Hold                38           146561.44
9         2005   In Process                41           135271.52
10        2005     Resolved                31            86549.12
11        2005     Disputed                14            61158.78

### Data Aggregation with Grouping

This query demonstrates how to use **grouping** and **aggregation** functions to analyze orders based on their **status** over multiple years. The key metrics, such as the **number of orders** and **total order amount**, provide valuable insights into customer behavior and business performance over time.